NIP-17 (old 24) Sealed Gift-Wrapped Messages for Private DMs and Small Group Chats (#686)
nostr-protocol/nips: 24.md

The use of other tags is highly encouraged (e.g. such as `content-warning` ([NIP-36](36.md)) if the content is NSFW). 

## 2. The Seal

A Seal is a `kind:13` event that authenticates the author of the encrypted message. It wraps the **unsigned** `kind:14` event (the Rumor) and is signed with the sender's regular key. 

```js
{
  "id": "<The usual hash>",
  "pubkey": "<Author's Main PubKey>",
  "content": "<XCHACHA-Encrypted Base64 of the Kind 14>",
  "kind": 13,
  "created_at": 1686840217,			// No need to match the current time
  "tags": [],   					// Always empty
  "sig": "<Author's Main PubKey Signature>"
}
```

The Sealed Rumor is **always** encrypted to a receiver's pubkey but **no** `p` tag pointing to the receiver is used. The receiver's information is not public. The only public information in this event is who is signing it. 

Clients MUST verify if pubkey of the `kind:13` is the pubkey on the `kind:14`, otherwise any author can impersonate others by simply changing the pubkey on `kind:14`.

> Clients SHOULD randomize `created_at` to reduce time-collisions.

Reviewer: Randomize is too vague. Over what range? Should i use times from 1972? Probably would be rejected by relays. Maybe "Clients SHOULD use a low-precision `created_at` that differs from <what?> to make time correlation of metadata less effective"

Author: It's hard to pin a range because it's dependent on how many GiftWraps are out there. Today Amethyst is randomizing within the past week. I believe Coracle does the same. But if we had lots of use, we wouldn't need to go that far.

Reviewer 2: This should probably just be removed, it's already in NIP 59 with more specificity.

Tags MUST always be empty. The inner event MUST always be unsigned. The encryption algorithm MUST use [NIP-44](44.md). 

## 3. The Gift Wrap

A GiftWrap event (`kind:1059`) wraps `kind:13` events. 

It is signed by a single-use (i.e. random) key and it is `p`-addressed to the receiver. 

The only public information is the receiver's public key. 

```js
{
  "id": "<The usual hash>",
  "pubkey": "<Random, one-time-use PubKey>",
  "content": "<XCHACHA-Encrypted Base64 of the Kind 13>",
  "kind": 1059,
  "created_at": 1686840217,
  "tags": [
    [ "p", "<Receiver>" ] // the only tag allowed
  ],
  "sig": "<Random, one-time-use PubKey Signature>"
}
```

A `p` tag for the recipient MUST be added to the `tags` field of a GiftWrap. 

> Clients SHOULD randomize `created_at` towards the past (some relays don't receive future events) to reduce time collisions.

Reviewer: See prior comment.

6. There is no moderation role (i.e. no group admins, no invitations or bans)
7. There is no chatroom secret that can leak or be mistakently shared
8. Messages can be fully recoverable in any client with the receiver or the sender's private key
9. Users and clients can opt-in for "Disappearing Messages" that are not recoverable with their private key
10. The protocol's messages can flow through public relays without loss of privacy. Private relays can increase privacy further, but they are not needed.
11. The protocol is extensible to make any other event kind fully private (private likes, private reports, private long-form content, etc)

The resulting protocol is a convergence of many past proposals, including [NIP-44](https://github.com/nostr-protocol/nips/pull/574), [NIP-59](https://github.com/nostr-protocol/nips/pull/468), [NIP-24](https://github.com/nostr-protocol/nips/pull/56), [NIP-103](https://github.com/nostr-protocol/nips/pull/499), [NIP-76](https://github.com/d-krause/nostr-nips/blob/nip76-draft-2/76.md). It brings their best ideas together in a slighly different approach. 

## Overview

> This protocol uses two main concepts to protect the transmission of a target event: Sealed Gossip Events and Gift Wraps ([NIP-59](https://github.com/nostr-protocol/nips/pull/468)). The target event is first **unsigned** (to make it unverifiable if it leaks -> "Gossip"), then Sealed, and then GiftWrapped.

Author: Correct. You can prove a pubkey is signing for "things". You can also prove another pubkey is receiving "things". But you don't know if they are messages, you don't know if the date is even relevant and you cannot prove the sender is sending to a given receiver without knowing one of the private keys.

  "created_at":1690665694,
  "id":"aa85f7c0dffd83a4474b3488fe443a53f75793c975a13fb25452798c7300ccf9",
  "kind":1059,
  "pubkey":"4f02eac59266002db5801adc5270700ca69d5b8f761d8732fab2fbf233c90cbd",
  "sig":"95b12fc477210748b447b3b7c8393a1b2ef5b1685122d0345534378d8ef8d2af4b05c1d403c12601d6ddc27e661e4a4452d6d0d676f38ff8a48feff8648d4b92",
  "tags":[
    [ "p", "2f27f7d08bd618c4c9a050d923ed915a693ecb3e5d399b26ef0829f29a51035f" ]
  ]
}
```
	
Notice that the signer is a random key.

#### 8.3.4. Broadcast Selectively
   
Broadcast the `kind:1059` event to the participant's relays only. Delete all the other events. 

## 9. Encryption Code Examples

### 9.1. JavaScript

> ```js
> // npm install @noble/curves @noble/ciphers @noble/hashes @scure/base @stablelib/xchacha20
> import {xchacha20} from '@noble/ciphers/chacha'
> import {secp256k1} from '@noble/curves/secp256k1'
> import {sha256} from '@noble/hashes/sha256'
> import {randomBytes} from '@noble/hashes/utils'
> import {base64} from '@scure/base'
> 
> 
> const utf8Decoder = new TextDecoder()
> const utf8Encoder = new TextEncoder()
> 
> export function getConversationKey(privkeyA: string, pubkeyB: string): Uint8Array {
>   const key = secp256k1.getSharedSecret(privkeyA, '02' + pubkeyB)
>   return sha256(key.subarray(1, 33))
> }
> 
> export function encrypt(
>   key: Uint8Array,
>   text: string,
>   ver = 1
> ): string {
>   if (ver !== 1) throw new Error('NIP44: unknown encryption version')
>   let nonce = randomBytes(24)
>   let plaintext = utf8Encoder.encode(text)
>   let ciphertext = xchacha20(key, nonce, plaintext, plaintext)
>   return `1,${base64.encode(nonce)},${base64.encode(ciphertext)}`
> }
> 
> export function decrypt(key: Uint8Array, data: string): string {
>   let dt = data.split(',')
>   if (dt.length !== 3) throw new Error('NIP44: unknown encryption version');
>   let v = Number.parseInt(dt[0])
>   if (v !== 1) throw new Error('NIP44: unknown encryption version')
> 
>   let nonce = base64.decode(dt[1])
>   let ciphertext = base64.decode(dt[2])
>   let plaintext = xchacha20(key, nonce, ciphertext)
>   let text = utf8Decoder.decode(plaintext)
>   return text
> }
> ```

Reviewer 2: ```suggestion
```js
import type {UnsignedEvent, Event} from 'nostr-tools'
import {getPublicKey, getEventHash, getSignature, generatePrivateKey} from "nostr-tools"
import {xchacha20} from '@noble/ciphers/chacha'
import {secp256k1} from '@noble/curves/secp256k1'
import {sha256} from '@noble/hashes/sha256'
import {randomBytes} from '@noble/hashes/utils'
import {base64} from '@scure/base'

export const fromHex = hex => Uint8Array.from(hex.match(/.{1,2}/g).map(byte => parseInt(byte, 16)))

export const now = (drift = 0) => {
  const t = Math.round(Date.now() / 1000)
  const factor = Math.pow(10, drift)
  const offset = factor - Math.round(factor * Math.random()) * 2

  return t + offset
}

export const utf8Decoder = new TextDecoder()

export const utf8Encoder = new TextEncoder()

export const getConversationKey = (privkeyA: string, pubkeyB: string): Uint8Array =>
  sha256(secp256k1.getSharedSecret(privkeyA, '02' + pubkeyB).subarray(1, 33))

export const getNonce = () => base64.encode(randomBytes(24))

export type Nip44Payload = {
  ciphertext: string
  nonce: string
  v: 1
}

export function encrypt(key: Uint8Array, text: string, v = 1): Nip44Payload {
  if (v !== 1) throw new Error('NIP44: unknown encryption version')

  const nonce = randomBytes(24)
  const plaintext = utf8Encoder.encode(text)
  const ciphertext = xchacha20(key, nonce, plaintext)

  return {
    ciphertext: base64.encode(ciphertext),
    nonce: base64.encode(nonce),
    v,
  }
}

export function decrypt(key: Uint8Array, data: Nip44Payload): string {
  if (data.v !== 1) throw new Error('NIP44: unknown encryption version');

  const nonce = base64.decode(data.nonce)
  const ciphertext = base64.decode(data.ciphertext)
  const plaintext = xchacha20(key, nonce, ciphertext)

  return utf8Decoder.decode(plaintext)
}

export const createRumor = (pubkey, {content, recipients, subject}) => {
  const rumor = {
    pubkey,
    content,
    kind: 14,
    created_at: now(),
    tags: recipients.map(k => ["p", k]).concat([["subject", subject]]),
  } as any

  rumor.id = getEventHash(rumor)

  return rumor as UnsignedEvent & {id: string}
}

export const createSeal = (privkey: string, pubkey, rumor) => {
  const sealKey = getConversationKey(privkey, pubkey)
  const content = JSON.stringify(encrypt(sealKey, JSON.stringify(rumor)))

  const seal = {
    content,
    kind: 13,
    created_at: now(5),
    pubkey: getPublicKey(privkey),
    tags: [],
  } as any

  seal.id = getEventHash(seal)
  seal.sig = getSignature(seal, privkey)

  return seal as Event
}

export const createGift = (pubkey, seal, privkey) => {
  const burnerKey = generatePrivateKey()
  const conversationKey = getConversationKey(burnerKey, pubkey)
  const content = JSON.stringify(encrypt(conversationKey, JSON.stringify(seal)))

  const gift = {
    content,
    kind: 1059,
    created_at: now(5),
    pubkey: getPublicKey(burnerKey),
    tags: [["p", pubkey]],
  } as any

  gift.id = getEventHash(gift)
  gift.sig = getSignature(gift, burnerKey)

  return gift as Event
}
```
```

Gift wraps alone are an effective tool to hide information from the public. However, the inner event can be easily re-broadcasted, breaking the privacy of the sender and leaking its metadata (e.g. author, event kind, dates, tags, etc). The GiftWrap MUST wrap a Sealed Gossip Event to make sure the receiver cannot break the privacy guarantees when re-broadcasting it. 

```js
{
  "id": "<The usual hash>",
  "pubkey": "<Random, one-time-use PubKey>",
  "content": "<XCHACHA-Encrypted Kind 13>",
  "kind": 1059,
  "created_at": 1686840217,
  "tags": [
    [ "p", "<Receiver>" ]
  ],
  "sig": "<Random, one-time-use PubKey Signature>"
}
```

A `p` tag for the recipient MUST be added to the `tags` field of a GiftWrap. This is how clients will download incoming "things". Other tags should not be added to this event. 

## 4. Encrypting with XChaCha (former NIP-44)

> To wrap an event for Sealed Gossip or GiftWrap, first serialize it to JSON. Then acquire a shared secret between the signing private key and the recipient's public key, a 24-byte nonce, and encrypt it with XChaCha (XChaCha is different than ChaCha). Convert both the nonce and the resulting byte array to Base64 and output an object with the following fields: 

Reviewer 3: maybe csv?

`1,base64(ciphertext),base64(nonce)` vs 
`{ "ciphertext": "<Base64-encoded ByteArray>", "nonce": "<Base64-encoded Nonce>", "v": <Version Code>  }`

Author: We already have a JSON parser in so many other places. Adding one more wouldn't change much. I don't think we have a CSV parser yet, so I would side with the JSON payload for now.

Reviewer 3: Which other things are using json?

Also, there's no need to conform to CSV spec -- just using basic `a,b,c` seems to work.

My concern here is not JSON itself, which is great, but the fact that it's nested JSON. E.g. JSON.parse(JSON.parse(response).content), which can be its own source of errors.

Reviewer 3: Also, using JSON would increase every message by 30 bytes. Base64 is only used because it saves a lot of traffic.

Readability is cool, but the fields are fixed for every algo version:

```
1,base64(nonce),base64(ciphertext) # always same
2,base64(nonce),base64(kdf_key),base64(ciphertext) # example algorithm 2, always same
```

Author: > Which other things are using nested json?

Reposts, Generic Reposts, and Community Post Approvals have the reposted event as a JSON string in the content. UserMetadata in Kind0 and the Relay list on Kind 3 is a JSON string. NIP28 Channel Metadata is also a JSON string. Private zaps use a JSON string to hide the private part. All NIP51 + Mute, Pin, Block lists have a private part which is a JSON string of the private tag list. Besides that, downloading Nostr events via WebSockets means parsing 1000s of JSONs a second. 

> JSON.parse(JSON.parse(response).content)

This is extremely common, though. 

> Using JSON would increase every message by 30 bytes. 

Not a big deal. Nostr encodes keys, hashes, and signatures in Hex strings, which is already wasteful. This protocol is not good for byte crunching. We should settle for what's the easiest encode/decode to facilitate adoption.

Reviewer 3: if it is already present in other events as you've mentioned, then I have no objections to using json

}

export function decrypt(key: Uint8Array, data: string): string {
  let dt = data.split(',')
  if (dt.length !== 3) throw new Error('NIP44: unknown encryption version');
  let v = Number.parseInt(dt[0])
  if (v !== 1) throw new Error('NIP44: unknown encryption version')

  let nonce = base64.decode(dt[1])
  let ciphertext = base64.decode(dt[2])
  let plaintext = xchacha20(key, nonce, ciphertext)
  let text = utf8Decoder.decode(plaintext)
  return text
}
```

### 9.2. Kotlin
  
```kotlin

import com.goterl.lazysodium.LazySodiumAndroid
import com.goterl.lazysodium.SodiumAndroid
import com.goterl.lazysodium.interfaces.AEAD
import com.goterl.lazysodium.utils.Base64MessageEncoder
import com.goterl.lazysodium.utils.Key
import java.util.Base64

data class EncryptedInfo(val ciphertext: String, val nonce: String, val v: Int)

enum class Nip44Version(val versionCode: Int) {
  Reserved(0),
  XChaCha20(1)
}

fun encryptXChaCha(msg: String, sharedSecret: ByteArray): EncryptedInfo {
  val lazySodium = LazySodiumAndroid(SodiumAndroid(), Base64MessageEncoder())

  val key = Key.fromBytes(sharedSecret)

  val nonce: ByteArray = lazySodium.nonce(24)

>   val cipherBase64: String = lazySodium.encrypt(msg, null, nonce, key, AEAD.Method.XCHACHA20_POLY1305_IETF)

Reviewer 3: Is that correct? `XCHACHA20_POLY1305`?

Author: Frankly, I am still trying to figure it out... I don't know if the JS version can be decrypted by the Kotlin version yet.

Reviewer 3: Use this: https://github.com/terl/lazysodium-java/blob/64dc78554f45eb9b81e74354d61eba4e3da2452b/src/main/java/com/goterl/lazysodium/SodiumJava.java#L139

Author: I don't have that method on Android. It's only available on desktop. :( 

Is it the only way?

Reviewer 3: i've thought the lazysodium-android library exposes all lazysodium methods? Are you 100% sure you can't access it?

Yes -- we need xchacha20, not xchacha20poly1305, they return different results.

Author: No, they are different implementations. The Android version doesn't have xchacha20XOR. The author said they were having lots of issues with that call in native android. :(

Reviewer 4: Please @vitorpamplona have you tried out this library: https://github.com/ionspin/kotlin-multiplatform-libsodium ? (The only issue is the experimental status of the lib :sweat_smile: )

Field `v` determines the algorithm version to be used. Current versions are: 
 - `0`: Reserved
 - `1`: XChaCha20
 
The stringified version of this object goes into the `.content` of the GiftWrap and Sealed Gossip events.  
 
## 5. Private Message Algorithm

To send a private message, create a kind `Kind:14` with the message in its `.content` and a p-Tag to the receiver. Other tags are encouraged to be added, such as `content-sensitivity` if the content is NSFW and `subject` to create a name to the conversation. 

Once `Kind:14` is created, remove the signature and serialize as JSON. Encrypt the serialized `Kind:14` to the receiver's public key and seal it with a `kind:13` event, signed by the sender's main key. 

For each receiver: 

1. Create a new key, serialize the seal, and encrypt it into a `kind:1059` with a tag to the receiver, signed by the new random key. 
2. Broadcast the `kind:1059` event to the relays of the receiver. 

For the sender: 

1. Create a new key, serialize the seal, and encrypt it into a `kind:1059` with a tag to the sender, signed by the new random key. 
2. Broadcast the `kind:1059` event to the relays of the sender. 

## 6. Private Group Chat Algorithm

> Private Group chats are `Kind:14`s for more than one `p` tag. There is no channel definition event. The set of authors + recipient keys defines a group.

Reviewer 2: So should the author tag himself?

Author: I merge the pubkeys in the p list with the author's pubkey so that it doesn't matter what the signing client does. :)

Author: We need to figure out how to remove `mentions` from being part of the group.

Reviewer 2: I remove the author's pubkey so it doesn't matter. I guess both approaches are compatible. And yes, that's a very good point. Maybe use "cc" tags instead? They don't need to be searchable.

Author: Either that or we go with a "mention" marker like on Kind1s.

Reviewer 2: That would work, but would be sort of error prone. The last thing you want is the person you're gossiping about to be invited to the conversation 😅

import {randomBytes} from '@noble/hashes/utils'
import {base64} from '@scure/base'

export const fromHex = hex => Uint8Array.from(hex.match(/.{1,2}/g).map(byte => parseInt(byte, 16)))

export const now = (drift = 0) => {
  const t = Math.round(Date.now() / 1000)
  const factor = Math.pow(10, drift)
  const offset = factor - Math.round(factor * Math.random()) * 2

  return t + offset
}

export const utf8Decoder = new TextDecoder()

export const utf8Encoder = new TextEncoder()

export const getConversationKey = (privkeyA: string, pubkeyB: string): Uint8Array =>
  sha256(secp256k1.getSharedSecret(privkeyA, '02' + pubkeyB).subarray(1, 33))

export const getNonce = () => base64.encode(randomBytes(24))

export type Nip44Payload = {
  ciphertext: string
  nonce: string
  v: 1
}

export function encrypt(key: Uint8Array, text: string, v = 1): Nip44Payload {
  if (v !== 1) throw new Error('NIP44: unknown encryption version')

  const nonce = randomBytes(24)
  const plaintext = utf8Encoder.encode(text)
  const ciphertext = xchacha20(key, nonce, plaintext)

  return {
    ciphertext: base64.encode(ciphertext),
    nonce: base64.encode(nonce),
    v,
  }
}

> export function decrypt(key: Uint8Array, data: Nip44Payload): string {
>   if (data.v !== 1) throw new Error('NIP44: unknown encryption version');
> 
>   const nonce = base64.decode(data.nonce)
>   const ciphertext = base64.decode(data.ciphertext)
>   const plaintext = xchacha20(key, nonce, ciphertext)
> 
>   return utf8Decoder.decode(plaintext)
> }
> 
> export const createRumor = (pubkey, {content, recipients, subject}) => {
>   const rumor = {
>     pubkey,
>     content,
>     kind: 14,
>     created_at: now(),
>     tags: recipients.map(k => ["p", k]).concat([["subject", subject]]),
>   } as any
> 
>   rumor.id = getEventHash(rumor)
> 
>   return rumor as UnsignedEvent & {id: string}
> }
> 
> export const createSeal = (privkey: string, pubkey, rumor) => {
>   const sealKey = getConversationKey(privkey, pubkey)
>   const content = JSON.stringify(encrypt(sealKey, JSON.stringify(rumor)))
> 
>   const seal = {
>     content,
>     kind: 13,
>     created_at: now(5),
>     pubkey: getPublicKey(privkey),
>     tags: [],
>   } as any
> 
>   seal.id = getEventHash(seal)
>   seal.sig = getSignature(seal, privkey)
> 
>   return seal as Event
> }
> 
> export const createGift = (pubkey, seal, privkey) => {
>   const burnerKey = generatePrivateKey()
>   const conversationKey = getConversationKey(burnerKey, pubkey)
>   const content = JSON.stringify(encrypt(conversationKey, JSON.stringify(seal)))
> 
>   const gift = {
>     content,
>     kind: 1059,
>     created_at: now(5),
>     pubkey: getPublicKey(burnerKey),
>     tags: [["p", pubkey]],
>   } as any
> 
>   gift.id = getEventHash(gift)
>   gift.sig = getSignature(gift, burnerKey)
> 
>   return gift as Event
> }

Reviewer 2: ```suggestion
import type {UnsignedEvent, Event} from "nostr-tools"
import {getPublicKey, getEventHash, getSignature} from "nostr-tools"
import {xchacha20} from "@noble/ciphers/chacha"
import {secp256k1} from "@noble/curves/secp256k1"
import {sha256} from "@noble/hashes/sha256"
import {randomBytes} from "@noble/hashes/utils"
import {base64} from "@scure/base"

export const utf8Decoder = new TextDecoder()

export const utf8Encoder = new TextEncoder()

export const fromHex = hex => Uint8Array.from(hex.match(/.{1,2}/g).map(byte => parseInt(byte, 16)))

export const toHex = bytes =>
  bytes.reduce((str, byte) => str + byte.toString(16).padStart(2, "0"), "")

export const now = (drift = 0) =>
  Math.round(Date.now() / 1000 - Math.random() * Math.pow(10, drift))

export const getConversationKey = (privkeyA: string, pubkeyB: string): Uint8Array =>
  sha256(secp256k1.getSharedSecret(privkeyA, "02" + pubkeyB).subarray(1, 33))

export const getNonce = () => base64.encode(randomBytes(24))

export type EncryptedPayload = {
  ciphertext: string
  nonce: string
  v: 1
}

export function encrypt(key: Uint8Array, text: string, v = 1): EncryptedPayload {
  if (v !== 1) throw new Error("NIP44: unknown encryption version")

  const nonce = randomBytes(24)
  const plaintext = utf8Encoder.encode(text)
  const ciphertext = xchacha20(key, nonce, plaintext)

  return {
    ciphertext: base64.encode(ciphertext),
    nonce: base64.encode(nonce),
    v,
  }
}

export function decrypt(key: Uint8Array, data: EncryptedPayload): string {
  if (data.v !== 1) throw new Error("NIP44: unknown encryption version")

  const nonce = base64.decode(data.nonce)
  const ciphertext = base64.decode(data.ciphertext)
  const plaintext = xchacha20(key, nonce, ciphertext)

  return utf8Decoder.decode(plaintext)
}

export const createRumor = event => {
  if (event.sig) {
    throw new Error("Rumor must not have a signature")
  }

  const rumor = {
    created_at: now(),
    content: "",
    tags: [],
    ...event,
  } as any

  rumor.id = getEventHash(rumor)

  return rumor as UnsignedEvent & {id: string}
}

export const createSeal = (authorPrivkey: string, recipientPubkey, rumor) => {
  const sealKey = getConversationKey(authorPrivkey, recipientPubkey)
  const content = JSON.stringify(encrypt(sealKey, JSON.stringify(rumor)))

  const seal = {
    content,
    kind: 13,
    created_at: now(5),
    pubkey: getPublicKey(authorPrivkey),
    tags: [],
  } as any

  seal.id = getEventHash(seal)
  seal.sig = getSignature(seal, authorPrivkey)

  return seal as Event
}

export const createWrap = (wrapperPrivkey, recipientPubkey, seal, tags = []) => {
  const conversationKey = getConversationKey(wrapperPrivkey, recipientPubkey)
  const content = JSON.stringify(encrypt(conversationKey, JSON.stringify(seal)))

  const wrap = {
    tags,
    content,
    kind: 1059,
    created_at: now(5),
    pubkey: getPublicKey(wrapperPrivkey),
  } as any

  wrap.id = getEventHash(wrap)
  wrap.sig = getSignature(wrap, wrapperPrivkey)

  return wrap as Event
}

export const wrap = (authorPrivkey, recipientPubkey, wrapperPrivkey, event, tags = []) => {
  const rumor = createRumor(event)
  const seal = createSeal(authorPrivkey, recipientPubkey, rumor)
  const wrap = createWrap(wrapperPrivkey, recipientPubkey, seal, tags)

  return wrap
}

export const unwrap = (recipientPrivkey, wrap) => {
  const sealKey = getConversationKey(recipientPrivkey, wrap.pubkey)
  const seal = JSON.parse(decrypt(sealKey, JSON.parse(wrap.content)))
  const rumorKey = getConversationKey(recipientPrivkey, seal.pubkey)
  const rumor = JSON.parse(decrypt(rumorKey, JSON.parse(seal.content)))

  return {wrap, seal, rumor}
}
```

Sealed Private Messages and Private Group Chats
-----------------------------------------------

`draft` `optional` `author:vitorpamplona` `author:kieran` `author:paulmillr`

This NIP creates a metadata-minimized encrypted messaging protocol for direct messages and closed group chats with the following privacy guarantees: 
1. Messages are encrypted to each participant's public key individually. 
2. Chat participant identities, each message's real date and time, event kinds, and any other tags are hidden from the public.
3. Senders and Receivers cannot be linked with public information alone, not even their random keys.
4. There is no central queue, channel or otherwise converging event id to link or count all messages in the same group. 
5. Minimal trust in counterparties: Counterparties cannot expose the details of your message, including the metadata, without exposing their entire profile (private key)
6. There is no moderation role (i.e. no group admins, no invitations or bans). 
7. Messages can be recoverable in any client with the receiver or the sender's private key
8. Users and clients can opt-in for Disappearing Messages that are not recoverable with the private key

> 9. The protocol's messages must flow through public relays. Private relays can increase privacy further, but they are not needed.

Reviewer 2: ```suggestion
9. The protocol's messages can flow through public relays without loss of privacy. Private relays can increase privacy further, but they are not needed.
```

10. The protocol becomes extensible to any other event kind to be transmitted privately by itself (private likes, private reports, private long-form-content, etc)

The resulting protocol is a convergence of many past proposals, including [NIP-44](https://github.com/nostr-protocol/nips/pull/574), [NIP-59](https://github.com/nostr-protocol/nips/pull/468), [NIP-24](https://github.com/nostr-protocol/nips/pull/56), [NIP-103](https://github.com/nostr-protocol/nips/pull/499), [NIP-76](https://github.com/d-krause/nostr-nips/blob/nip76-draft-2/76.md). It brings their best ideas together is a slighly different approach. 

## Overview

> This protocol uses two main concepts to protect the transmission of a target event: Sealed Gossip Events and Gift Wraps ([NIP-59](https://github.com/nostr-protocol/nips/pull/468)). The target event is first **unsigned** (to make it unverifiable if it leaks -> "Gossip"), then Sealed, and then GiftWrapped.

Reviewer 2: Gossip doesn't seem quite right, what about "Rumor"?

Author: Yeah, I am struggling with these names as well. :(

Author: Maybe we can change from Sealed Gossip to Ratifed Rumor. It's strange to Seal a Rumor.

Reviewer 2: I don't mind that, I like the Sealed metaphor too. A letter with no signature means nothing, but if you put it in an envelope with a wax seal it's authenticated. You wouldn't ratify a rumor either 😂

Reviewer 2: Could go with metonymy and just call the outer event a "Seal" since that's effectively the role it's performing

Reviewer 2: In my implementation I ended up doing "rumor/seal/wrapper"

As an early example: this is the flow of a private message: 

1. Target Event   `To Receiver: "Hi Receiver, did you finish the transfer? -- Signed, Sender"`
2. Gossip         `To Receiver: "Hi Receiver, did you finish the transfer? -- From Sender" (unsigned)`
3. Sealed Gossip  `<ENCRYPTED Gossip> -- Signed, Sender`
4. Gift Wrap      `To Receiver: <ENCRYPTED Sealed Gossip> -- Signed, Anon`

The Gift Wrap is the only event sent to relays. 

Any event kind can be made a Gossip by removing the signature. 

## 1. A new ChatMessage Kind as Gossip

This NIP defines `Kind:14` as a Chat Message in a similar way `Kind:4` does, but it's never signed and if it is signed, the information in it is **fully public**. This event requires the rest of the protocol to make it private. 

Direct Message Payload: 
```js
{
  "id": "<The usual hash>",
  "pubkey": "<author key in hex>",
  "content": "This is a DM",
  "kind": 14,
  "created_at": 1686840217,
  "tags": [ <one or more receivers, mentions, citations, additional fields like content-sensitivity, etc> ],
}
```

## 2. The Sealed Gossip Event Kind

> A Sealed Gossip is a `kind:13` event that wraps an unsigned event (the Gossip) with the sender's regular key. The Sealed Gossip is **always** encrypted to a receiver's pubkey but there are no `p-Tag` associated with this event. There is no way to know who the Gossip is for without the receiver or the sender's private key. The only public information in this event is who is signing it. 

Reviewer 2: ```suggestion
A Sealed Gossip is a `kind:13` event that wraps an unsigned event (the Gossip) with the sender's regular key. The Sealed Gossip is **always** encrypted to a receiver's pubkey but there is no `p` tag associated with this event. There is no way to know who the Gossip is for without the receiver or the sender's private key. The only public information in this event is who is signing it. 
```

```js
{
  "id": "<The usual hash>",
  "pubkey": "<Real Author's PubKey>",
  "content": "<XCHACHA-Encrypted Gossip Kind>",
  "kind": 13,
  "created_at": 1686840217,
  "tags": [],
  "sig": "<Real Author's PubKey Signature>"
}
```

Tags MUST must always be empty in a `kind:13`, 

The inner event MUST always be unsigned 

> The enryption algorithm MUST use a versioned payload with a XChaCha chiper. 

Reviewer 2: ```suggestion
The enryption algorithm MUST use a versioned payload with a XChaCha cipher. 
```

```js
{
  "id": "<The usual hash>",
  "pubkey": "<Random, one-time-use PubKey>",
  "content": "<XCHACHA-Encrypted Kind 13>",
  "kind": 1059,
  "created_at": 1686840217,
  "tags": [
    [ "p", "<Receiver>" ]
  ],
  "sig": "<Random, one-time-use PubKey Signature>"
}
```

A `p` tag for the recipient MUST be added to the `tags` field of a GiftWrap. This is how clients will download incoming "things". Other tags should not be added to this event. 

## 4 Encrypting with XChaCha (former NIP-44)

To wrap an event for Sealed Gossip or GiftWrap, first serialize it to JSON. Then acquire a shared secret between the signing private key and the recipient's public key, a 24-byte nonce, and encrypt it with XChaCha (XChaCha is different than ChaCha). Convert both the nonce and the resulting byte array to Base64 and output an object with the following fields: 

```js
{
  ciphertext: <Base64-encoded ByteArray>
  nonce: <Base64-encoded Nonce>
  v: <Algorithm Version Integer>  
}
```

Field `v` determines the algorithm version to be used. Current versions are: 
 - `0`: Reserved
 - `1`: XChaCha20
 
## 5. Private Message Algorithm

> To send a private message, create a kind `Kind:14` with the message in it's `.content` and a p-Tag to the receiver. Other tags are encouraged to be added, such as `content-sensitivity` if the content is NSFW and `subject` to create a name to the conversation. 

Reviewer 2: ```suggestion
To send a private message, create a kind `Kind:14` with the message in its `.content` and a p-Tag to the receiver. Other tags are encouraged to be added, such as `content-sensitivity` if the content is NSFW and `subject` to create a name to the conversation. 
```

Once `Kind:14` is created, remove the signature and serialize as JSON. Encrypt the serialized `Kind:14` to the receiver's public key and seal it with a `kind:13` event, signed by the sender's main key. 

For each receiver: 

1. Create a new key, serialize the seal, and encrypt it into a `kind:1059` with a tag to the receiver, signed by the new random key. 
2. Broadcast the `kind:1059` event to the relays of the receiver. 

For the sender: 

1. Create a new key, serialize the seal, and encrypt it into a `kind:1059` with a tag to the sender, signed by the new random key. 
2. Broadcast the `kind:1059` event to the relays of the sender. 

## 6. Private Group Chat Algorithm

> Private Group chats are `Kind:14`s for more than one `p` tag. There is no channel definition event. The group of authors + p' tags is a group. 

Reviewer 2: ```suggestion
Private Group chats are `Kind:14`s for more than one `p` tag. There is no channel definition event. The set of authors + recipient keys defines a group.
```

To send a private message to a group, create a kind `Kind:14` event with the message in its `.content` and add a p-Tag to each receiver. Other tags are encouraged to be added, such as `content-sensitivity` if the content is NSFW and `subject` to create a name to the conversation. Once `Kind:14` is created, remove the signature and serialize it as JSON.

For each receiver: 

1. Encrypt the serialized `Kind:14` to the **receiver's public key** and seal it with a `kind:13` event, signed by the sender's main key. 
2. Create a new key, serialize the seal, and encrypt it into a `kind:1059` with a tag to the **receiver**, signed by the new random key. 
3. Broadcast the `kind:1059` event to the relays of the **receiver**. 

For the sender: 

1. Create a new key, serialize the seal, and encrypt it into a `kind:1059` with a tag to the **sender**, signed by the new random key. 
2. Broadcast the `kind:1059` event to the relays of the **sender**. 

> Clients SHOULD render every different group of (author + p' tags) as a group with continuing conversations. If a new p tag is added, a new group is created. 

Reviewer 2: ```suggestion
Clients SHOULD render every distinct set of author + recipient pubkeys as a group with continuing conversations. If a new p tag is added, a new group is created. 
```

The `subject` tag defines any name for the group or the current topic of the discussion. Any member of the group can change the subject. 

## 7. Final Considerations

This proposal makes heavy use of one-time keys to elevate the privacy design of Nostr messages. It can be applied to any Nostr event kind (e.g. private reactions), but the focus here is on private messages. For instance, the use of a Gossip Event enhances a paid-subscriber-only event kind by making it difficult (not impossible) to leak the paid content in a publicly verifiable payload. 

> It's worth noting that the only created_at time that matters is with the Gossip event. All the other time-attributes can be tweaked to reduce the time-collision of send and receive Gift Wraps and Sealed Events.

Reviewer 2: For the purposes of balancing retrieval and privacy, maybe we could recommend how much to randomize the timestamp. Now +/- one day? One week?

Reviewer 2: ```suggestion
It's worth noting that the only created_at time that matters is with the Gossip event. All the other time-attributes SHOULD be tweaked to reduce the time-collision of send and receive Gift Wraps and Sealed Events.

Relays may choose not to store gift wrapped events due to them not being publicly useful. Clients MAY choose to attach a certain amount of proof-of-work to the wrapper event per NIP-13 in a bid to demonstrate that the event is not spam or a denial-of-service attack.

To protect recipient metadata, relays SHOULD only serve kind 1059 events intended for the marked recipient. When possible, clients should only send wrapped events to read relays for the recipient that implement AUTH, and refuses to serve wrapped events to non-recipients.
```

  "created_at":1690558348,
  "id":"96b5def1d84c5c33d90fc3a36133a32eba0eca98dd67f630d442a2456d5b9c4e",
  "kind":1059,
  "pubkey":"1c1b8ab402dcdf7bdaeba8cf5056b2ec5d2958886bcaf342cd37d71cfde1bf8d",
  "sig":"1144271fc0e3d07df0ba72b58a4544a469e83314bf29dc9275ea67d0d0a7ec4989487fdcf1ae9958710e0956d56a3833e92ce40dd22f0de628718888538d098f",
  "tags":[
    [ "p", "2d5ad8fb75c30aeb153e6d8b6a7275424be54d196f4837d658cdd36c81dad8d3" ]
  ]
}
```

5.3 Broadcast the `kind:1059` to the sender's relays

## 10. Encryption Code Examples

### 10.1 JavaScript

```js

> // npm install @noble/curves @noble/hashes @scure/base @stablelib/xchacha20

Reviewer 2: ```suggestion
// npm install @noble/curves @noble/ciphers @noble/hashes @scure/base @stablelib/xchacha20
```

import {xchacha20} from '@noble/ciphers/chacha'
import {secp256k1} from '@noble/curves/secp256k1'
import {sha256} from '@noble/hashes/sha256'
import {randomBytes} from '@noble/hashes/utils'
import {base64} from '@scure/base'

> import {utf8Decoder, utf8Encoder} from './utils.ts'

Reviewer 2: ```suggestion

const utf8Decoder = new TextDecoder()
const utf8Encoder = new TextEncoder()
```

## Overview

This protocol uses two main concepts to protect the transmission of a target event: Sealed Gossip Events and Gift Wraps ([NIP-59](https://github.com/nostr-protocol/nips/pull/468)). The target event is first **unsigned** (to make it unverifiable if it leaks -> "Gossip"), then Sealed, and then GiftWrapped.

As an early example: this is the flow of a private message: 

1. Target Event   `To Receiver: "Hi Receiver, did you finish the transfer? -- Signed, Sender"`
2. Gossip         `To Receiver: "Hi Receiver, did you finish the transfer? -- From Sender" (unsigned)`
3. Sealed Gossip  `<ENCRYPTED Gossip> -- Signed, Sender`
4. Gift Wrap      `To Receiver: <ENCRYPTED Sealed Gossip> -- Signed, Anon`

The Gift Wrap is the only event sent to relays. 

Any event kind can be made a Gossip by removing the signature. 

## 1. A new ChatMessage Kind as Gossip

> This NIP defines `Kind:14` as a Chat Message in a similar way `Kind:4` does, but it's never signed and if it is signed, the information in it is **fully public**. This event requires the rest of the protocol to make it private. 

Reviewer 2: ```suggestion
This NIP defines `Kind:14` as a Chat Message in a similar way `Kind:4` does, but it's never signed and if it is signed, the information in it is **fully public**. This event requires the rest of the protocol to make it private. Any other kind may be wrapped in the same way as a kind 14 following the steps described below in order to implement any other kind of private data transfer.
```

## 8. P2P Message Example, Step By Step

Let's send a message from `0956f4d1f65968e303f22220382550ef5979de1a4f0f136023c947d21a344bf6` to `7185de5d1f6ca0787b689738bc6895e88f88781e4e1961400742fbfbc19c6863`, saying "Hi There!"

1. Create a `kind:14` with the message, the receiver, and all tags you want. 

```json
{
  "content":"Hi There!",
  "created_at":1690556556,
  "id":"76375b8c43669f3f412333a326c1001f371726faf4acb74bf2f7642a1d2923c1",
  "kind":14,
  "pubkey":"0956f4d1f65968e303f22220382550ef5979de1a4f0f136023c947d21a344bf6",
  "sig":"80aae4ba812e55586bdc1c2dcd3ff7b680b19b9ca7e019465a74819508cc44008e3cce99320c9d6256575f8a32c1873831cce8047b934e17ea442cd4be4ed072",
  "tags":[
    [ "p", "7185de5d1f6ca0787b689738bc6895e88f88781e4e1961400742fbfbc19c6863" ]
  ]
}
```

2. Remove the signature to create an unverifiable Gossip

```json
{
  "content":"Hi There!",
  "created_at":1690556556,
  "id":"76375b8c43669f3f412333a326c1001f371726faf4acb74bf2f7642a1d2923c1",
  "kind":14,
  "pubkey":"0956f4d1f65968e303f22220382550ef5979de1a4f0f136023c947d21a344bf6",
  "tags":[
    [ "p", "7185de5d1f6ca0787b689738bc6895e88f88781e4e1961400742fbfbc19c6863" ]
  ]
}
```

> 3. Encrypt the JSON-Stringified Gossip the Sender's key

Reviewer 2: ```suggestion
3. Encrypt the JSON-Stringified Gossip with the Sender's key
```

  "created_at":1690558348,
  "id":"820609ba08d8a53d39ef7ceb297099999fd76107907d8cf2d62bf6b356b2ca8b",
  "kind":1059,
  "pubkey":"c632580bb8d46b65ca0bc58bfd78485bf1378ec494735f8b7002439fd59e443a",
  "sig":"906b82e1ea5290e01fa5b1e69d86140ef58e3649576233fc4ba15ad73c934ca75ec1ca8270d2000a4dba2a562ed1ae6bf2554705eb6f9191234acf8479d21010",
  "tags":[
    [ "p", "063491fa5da40f7766802ef33311428332754372999a6f8411df2420fc7a5f01" ]
  ]
}

> ````

Reviewer 2: Too many backticks
